### PR TITLE
Benchmark fine-tuned wav2vec2 on IEMOCAP

### PR DESCRIPTION
**Pull request by BenoitWang**
Hi everyone, I am working on a benchmarking of wav2vec2.0 on Speech Emotion Recognition with IEMOCAP. The result shows a 75.6% test accuracy, with only an average time pooling and a simple classifier.

### CODE REVIEW

**Comment by mravanelli** [comment]
Pretty interesting @BenoitWang! Let me try to replicate this number and then we can eventually merge!

**Comment by aheba** [comment]
hi Yingzhi, thanks for the PR,

Let replicates the numbers and review the code, 

Best

File: recipes/IEMOCAP/emotion_recognition/hparams/train_with_wav2vec2.yaml [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+experiment: emotion_recognition
+# Set up folders for reading from and writing to
+# Dataset will be downloaded to the `data_folder`
+data_folder: !PLACEHOLDER # e.g., /storage/IEMOCAP/IEMOCAP_ahsn_leave-two-speaker-out
+output_folder: !ref results/wav2vec/<seed>
+save_folder: !ref <output_folder>/save
+train_log: !ref <output_folder>/train_log.txt
+
+# URL for the wav2vec2 model, you can change to benchmark diffrenet models
+wav2vec2_hub: "facebook/wav2vec2-base-960h"
```
Comment (aheba) [inline recipes/IEMOCAP/emotion_recognition/hparams/train_with_wav2vec2.yaml:19]: why are you using the non fine tuned one on ASR ?
did you see the difference

File: recipes/IEMOCAP/emotion_recognition/hparams/train_with_wav2vec2.yaml [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+
+# The train logger writes training statistics to a file, as well as stdout.
+train_logger: !new:speechbrain.utils.train_logger.FileTrainLogger
+    save_file: !ref <train_log>
+
+ckpt_interval_minutes: 15 # save checkpoint every N min
+
+# Training parameters
+number_of_epochs: 30
+batch_size: 4
```
Comment (aheba) [inline recipes/IEMOCAP/emotion_recognition/hparams/train_with_wav2vec2.yaml:38]: batch of 4 only with GPU of V100 32GB ?

File: recipes/IEMOCAP/emotion_recognition/hparams/train_with_wav2vec2.yaml [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+lr: 0.0001
+lr_wav2vec2: 0.00001
+
+#freeze all wav2vec2
+freeze_wav2vec2: False
+#set to true to freeze the CONV part of the wav2vec2 model
+freeze_wav2vec2_conv: True
```
Comment (aheba) [inline recipes/IEMOCAP/emotion_recognition/hparams/train_with_wav2vec2.yaml:45]: @BenoitWang , do you try unfreezing conv layer of wav2vec ?
you got bad number ?

File: recipes/IEMOCAP/emotion_recognition/train_with_wav2vev2.py [hunk 4]
```diff
@@ -0,0 +1,297 @@
...
+class EmoIdBrain(sb.Brain):
+    def compute_forward(self, batch, stage):
+        """Computation pipeline based on a encoder + emotion classifier.
+        """
+        batch = batch.to(self.device)
+        wavs, lens = batch.sig
+
+        wav2vec2_out = self.modules.wav2vec2(wavs)
+
+        average_pooling = sb.nnet.pooling.Pooling1d(
```
Comment (aheba) [inline recipes/IEMOCAP/emotion_recognition/train_with_wav2vev2.py:28]: Yingzhi, Could you please use:
AdaptativeAvgPool1D instead of sb.nnet.pooling.Pooling1d

you can do it in that way:
create you obj in the yaml:
```
avg_pool: !new:torch.nn.AdaptiveAvgPool1d
              output_size:1
```
and in you python file
```
# last dim will be used for AdaptativeAVG pool
input = input.permute(0,2,1)
out = this.avg_pool(input)
out.view(out.shape[0],-1)
```

File: recipes/IEMOCAP/emotion_recognition/train_with_wav2vev2.py [hunk 5]
```diff
@@ -0,0 +1,297 @@
...
+        )
+
+        outputs = self.modules.output_mlp(averaged_out)
+        outputs = self.hparams.log_softmax(outputs)
+        return outputs
+
+    def compute_objectives(self, predictions, batch, stage):
+        """Computes the loss using speaker-id as label.
+        """
+        emoid, _ = batch.to(self.device).emo_encoded
```
Comment (aheba) [inline recipes/IEMOCAP/emotion_recognition/train_with_wav2vev2.py:46]: Why loading the batch with the specific device in your data_io function ?

**Review by aheba — CHANGES_REQUESTED** [review]

**Comment by aheba** [comment]
Hey @BenoitWang , thanks for this PR, I think you should use AdaptativeAVGPOOL1D instead of creating for each forward a new pool1D.

https://pytorch.org/docs/stable/generated/torch.nn.AdaptiveAvgPool1d.html

feel free to apply the changes,

I'm running an exp to see if I can reproduce your performance

**Comment by mravanelli** [comment]
Thank you very much for this PR. Is everything ok with the data preparation? My student @pierreyanni told me that the one in our current recipe assumes a folder structure different from those of the original dataset. Did you see issues on that?

**Comment by BenoitWang** [comment]
Hi @mravanelli , thanks for the interest. Yes, there is a different folder srtructure, and I used the same script as in the PASE project for restructuring the IEMOCAP folder. I tested the existing ecapa-tdnn on it and I got 66% test accuracy.

**Comment by BenoitWang** [comment]
Hi @aheba , thanks for the comments. I have modified the files according to your comments, and I am now carring out more experiments to see the differences of what you said

**Comment by mravanelli** [comment]
Maybe it makes sense to take this opportunity to change the data preparation as well and make it compliant with the the original folder. What do you think?

**Comment by aheba** [comment]
@BenoitWang , Could you handle it ?

From my side, I'm trying to reproduce the numbers..

**Comment by BenoitWang** [comment]
Yes @mravanelli it makes no sense to run two different preparations and I am going to work on a new preparation script.

File: recipes/IEMOCAP/emotion_recognition/hparams/train_with_wav2vec2.yaml [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+
+# Model parameters
+Wav2vec_encoder_dim: 768
```
[the remark this answers is not in the document]
Comment (BenoitWang) [reply recipes/IEMOCAP/emotion_recognition/hparams/train_with_wav2vec2.yaml]: Done

File: recipes/IEMOCAP/emotion_recognition/train_with_wav2vev2.py (hunk 4 above)
Comment (BenoitWang) [reply recipes/IEMOCAP/emotion_recognition/train_with_wav2vev2.py]: Done

File: recipes/IEMOCAP/emotion_recognition/hparams/train_with_wav2vec2.yaml (hunk 2 above)
Comment (BenoitWang) [reply recipes/IEMOCAP/emotion_recognition/hparams/train_with_wav2vec2.yaml:38]: Yes, got OOM with batch_size 8

File: recipes/IEMOCAP/emotion_recognition/hparams/train_with_wav2vec2.yaml (hunk 1 above)
Comment (BenoitWang) [reply recipes/IEMOCAP/emotion_recognition/hparams/train_with_wav2vec2.yaml:19]: yes you are right I am carrying out more experiments to be precise

File: recipes/IEMOCAP/emotion_recognition/hparams/train_with_wav2vec2.yaml (hunk 3 above)
Comment (BenoitWang) [reply recipes/IEMOCAP/emotion_recognition/hparams/train_with_wav2vec2.yaml:45]: yes unfreeze conv shows a little bit worse but I am carring out more tests to be precise

File: recipes/IEMOCAP/emotion_recognition/train_with_wav2vev2.py (hunk 5 above)
Comment (BenoitWang) [reply recipes/IEMOCAP/emotion_recognition/train_with_wav2vev2.py]: Done

**Comment by aheba** [comment]
@BenoitWang, I'm running an exp to reproduce your results but the `valid error_rate` is very high :
```
Epoch: 1, lr: 1.00e-04, wave2vec_lr: 1.00e-05 - train loss: 1.20 - valid loss: 1.02, valid error_rate: 4.39e-01
Epoch: 2, lr: 1.00e-04, wave2vec_lr: 1.00e-05 - train loss: 9.64e-01 - valid loss: 9.13e-01, valid error_rate: 3.71e-01
Epoch: 3, lr: 1.00e-04, wave2vec_lr: 1.00e-05 - train loss: 8.48e-01 - valid loss: 8.56e-01, valid error_rate: 3.38e-01
Epoch: 4, lr: 1.00e-04, wave2vec_lr: 1.00e-05 - train loss: 7.55e-01 - valid loss: 9.17e-01, valid error_rate: 2.98e-01
Epoch: 5, lr: 1.00e-04, wave2vec_lr: 1.00e-05 - train loss: 6.66e-01 - valid loss: 9.52e-01, valid error_rate: 3.18e-01
Epoch: 6, lr: 9.00e-05, wave2vec_lr: 9.00e-06 - train loss: 5.86e-01 - valid loss: 8.97e-01, valid error_rate: 2.84e-01
Epoch: 7, lr: 9.00e-05, wave2vec_lr: 9.00e-06 - train loss: 5.24e-01 - valid loss: 1.31, valid error_rate: 3.00e-01
Epoch: 8, lr: 8.10e-05, wave2vec_lr: 8.10e-06 - train loss: 4.61e-01 - valid loss: 1.09, valid error_rate: 2.77e-01
Epoch: 9, lr: 8.10e-05, wave2vec_lr: 8.10e-06 - train loss: 4.25e-01 - valid loss: 1.34, valid error_rate: 2.77e-01
Epoch: 10, lr: 7.29e-05, wave2vec_lr: 7.29e-06 - train loss: 3.96e-01 - valid loss: 1.19, valid error_rate: 2.57e-01
Epoch: 11, lr: 7.29e-05, wave2vec_lr: 7.29e-06 - train loss: 3.55e-01 - valid loss: 1.40, valid error_rate: 2.60e-01
Epoch: 12, lr: 6.56e-05, wave2vec_lr: 6.56e-06 - train loss: 3.31e-01 - valid loss: 1.44, valid error_rate: 2.64e-01
Epoch: 13, lr: 5.90e-05, wave2vec_lr: 5.90e-06 - train loss: 2.76e-01 - valid loss: 1.53, valid error_rate: 2.55e-01
Epoch: 14, lr: 5.90e-05, wave2vec_lr: 5.90e-06 - train loss: 2.64e-01 - valid loss: 1.48, valid error_rate: 2.50e-01
```
Could you please share with us your training log ?

**Comment by aheba** [comment]
hey @BenoitWang , @mravanelli ,
 
I notice that there is a difference between @BenoitWang  prepare and the one from https://github.com/eesungkim/Speech_Emotion_Recognition_DNN-ELM

@BenoitWang  prepare have:
```
==================================================
Total Session_1 : 1085
{'neu': 384, 'hap': 135, 'sad': 194, 'ang': 229, 'sur': 0, 'fea': 0, 'dis': 0, 'fru': 0, 'exc': 143, 'xxx': 0}
==================================================

==================================================
Total Session_2 : 1023
{'neu': 362, 'hap': 117, 'sad': 197, 'ang': 137, 'sur': 0, 'fea': 0, 'dis': 0, 'fru': 0, 'exc': 210, 'xxx': 0}
==================================================

==================================================
Total Session_3 : 1151
{'neu': 320, 'hap': 135, 'sad': 305, 'ang': 240, 'sur': 0, 'fea': 0, 'dis': 0, 'fru': 0, 'exc': 151, 'xxx': 0}
==================================================

==================================================
Total Session_4 : 1031
{'neu': 258, 'hap': 65, 'sad': 143, 'ang': 327, 'sur': 0, 'fea': 0, 'dis': 0, 'fru': 0, 'exc': 238, 'xxx': 0}
==================================================

==================================================
Total Session_5 : 1241
{'neu': 384, 'hap': 143, 'sad': 245, 'ang': 170, 'sur': 0, 'fea': 0, 'dis': 0, 'fru': 0, 'exc': 299, 'xxx': 0}
==================================================

```

when using original prepare:
```
iemocap_prepare - ==================================================
iemocap_prepare - Total Session_1 : 564
iemocap_prepare - {'neu': 161, 'hap': 102, 'sad': 90, 'ang': 167, 'sur': 0, 'fea': 0, 'dis': 0, 'fru': 0, 'exc': 44, 'xxx': 0}
iemocap_prepare - ==================================================

iemocap_prepare - ==================================================
iemocap_prepare - Total Session_2 : 493
iemocap_prepare - {'neu': 145, 'hap': 63, 'sad': 97, 'ang': 115, 'sur': 0, 'fea': 0, 'dis': 0, 'fru': 0, 'exc': 73, 'xxx': 0}
iemocap_prepare - ==================================================

iemocap_prepare - ==================================================
iemocap_prepare - Total Session_3 : 524
iemocap_prepare - {'neu': 122, 'hap': 75, 'sad': 115, 'ang': 150, 'sur': 0, 'fea': 0, 'dis': 0, 'fru': 0, 'exc': 62, 'xxx': 0}
iemocap_prepare - ==================================================

iemocap_prepare - ==================================================
iemocap_prepare - Total Session_4 : 497
iemocap_prepare - {'neu': 84, 'hap': 34, 'sad': 62, 'ang': 243, 'sur': 0, 'fea': 0, 'dis': 0, 'fru': 0, 'exc': 74, 'xxx': 0}
iemocap_prepare - ==================================================


iemocap_prepare - ==================================================
iemocap_prepare - Total Session_5 : 510
iemocap_prepare - {'neu': 97, 'hap': 37, 'sad': 112, 'ang': 139, 'sur': 0, 'fea': 0, 'dis': 0, 'fru': 0, 'exc': 125, 'xxx': 0}
iemocap_prepare - ==================================================
```

@mravanelli , could you double check that the transformed data have 5531 wav file rather than 2588 ?

**Comment by aheba** [comment]
I found the error, a missing condition in the git of IEMOCAP data transform.

I'm fixing it...

**Comment by aheba** [comment]
Hi @BenoitWang , thank you for this PR,

Let me summarise what's done in this PR:
- [x] data transformation from orignal to structured IEMOCAP data
- [x] add wav2vec2 encoder. Freezing CNN allow us to gain 2% of perf. Using the non fine tuned wav2vec2 instead of wav2vec2 fine tuned on Librispeech allow us to gain 4% (it definitely matters for non verbal information).
- [x] we tried avgpooling, attentivestatisticalpooling and seems give same performance.

We are exploring `Focal Loss` because of the unbalence emotion classes: neutral: 30.7%, happy: 30.32%, sad: 19.04%, anger: 19.94%   (neutral represent around 35% of the corpus)

@BenoitWang propose to add weighted evaluation because of the unbalance classes

I think we should add result file to see good/worth prediction for deeper analysis... what do you think @mravanelli

**Comment by BenoitWang** [comment]
@aheba thanks a lot for all this work Abdel. And for the metric, I don't think we have now unweighted accuracy (UA) in our actual version, which is more representative towards unbalanced data. Do you think that we add UA taking this chance @mravanelli ?

**Comment by mravanelli** [comment]
Sure, we can add it. As far as I remember, most of the papers on IEMOCAP
report the standard accuracy (but I might be wrong, please double check).

**Comment by aheba** [comment]
@BenoitWang , let's add:
- [ ] add confusion matrix output (@aheba )
- [ ] add unweighted perf stat (@BenoitWang)
- [ ] add focal loss (@aheba)
- [ ] deploy our best model into HF + drive (@aheba)

**Comment by mravanelli** [comment]
Hi @BenoitWang and @aheba,
thank you for this PR. Let me share with you some comments:

1. adaptive average pooling can be critical because it includes zero-padding elements. I suggest using our statistical pooling. If you don't need the standard deviation, we can add an option in `speechbrain.nnet.pooling.StatisticsPooling` to remove it.

2. `wav2vec2_hub: "facebook/wav2vec2-base"` => I suggest trying the voxpopuli version as well (`facebook/wav2vec2-large-100k-voxpopuli`)

3. fix typo `train_with_wav2vev2.py` => `fix typo train_with_wav2vec2.py`

**Comment by mravanelli** [comment]
The code looks good to me. I still have to try it, but I have the following minor comments on the README file:

1. `"Get the IEMOCAP dataset from https://sail.usc.edu/iemocap/iemocap_release.htm and put it in the same folder as iemocap_prepare.py under the name IEMOCAP_processed.tar.gz"` => I think this is outdated, right? If so we have to remove it.

2. `Gdrive download` => Please, add the link

3. "`Note on Data Preparation`" => I think this is outdated, right?

4. Add the HuggigFace repository + easy inference instructions (also add the link and instructions in the README file)

**Comment by aheba** [comment]
Hi @TParcollet,

I'm reporting Hubert perf on IEMOCAP which is very good (AVG: 80%). In the meanwhile, I do small changes for HF interface, where I add model/conf dict for loading HF models,

is it fine for you ?
```
# We check if transformers is installed.
try:
    from transformers import Wav2Vec2Model, HubertModel
    from transformers import Wav2Vec2Config, HubertConfig
    from transformers import Wav2Vec2FeatureExtractor
except ImportError:
    print(
        "Please install transformer from HuggingFace to use wav2vec2/Hubert !"
    )

HF_models = {"wav2vec2": Wav2Vec2Model, "hubert": HubertModel}

HF_config = {"wav2vec2": Wav2Vec2Config, "hubert": HubertConfig}
```

and:
```
        if "hubert" in source:
            config = HF_config.get("hubert")
            model = HF_models.get("hubert")
        else:
            config = HF_config.get("wav2vec2")
            model = HF_models.get("wav2vec2")
```

**Comment by TParcollet** [comment]
@aheba I propose that we simply rename the huggingface_wav2vec2.py into huggingface_models.py something like that. Be careful that this will induce a refactoring of all recipes using HF models. After this change, LGTM.

**Comment by aheba** [comment]
hi @mravanelli ,

All the modification is done, I'm waiting for @nauman-daw  and @hwidong-na  answer about the AVG pooling index + additive gaussian noise..

here is the HF repos https://huggingface.co/speechbrain/emotion-recognition-wav2vec2

We are ok for merging this PR.

Thank you @BenoitWang  for your effort ;)
